### DISCUSSION TITLE
Add options for monika to run agent in localhost or remote

### DISCUSSION THREAD

**Discussion by ilmiawan** [hyperjumptech/monika#153] [Ideas]
**Describe the bug**
Currently monika only able to run from local agent, we want monika to also run from remote agent as in the symon original idea.
By running remote agent we can try multiple host or region for monitoring.

**To Reproduce**
1. Go to Monika
2. Add configuration
3. Run Monika

Currently the monitoring from Monika is only run from the local agent, checking the url set in probe. 

**Expected behavior**
Monika can run the checking to the url from agent deployed in remote servers.

**Comment by newm4n** [comment]
I have reconfirmed this issue with Pak @ariya , and my confirmation has a firm my stand on how we should do this technically. 
The explanation on how Monika should make a HTTP request call remotely is not by installing another monika instance remotely, making a monika to monika communication where one monika instruct another monika to make HTTP call. Instead, we are utilizing SSH communication to spawn CURL process on remote machine to do the HTTP call.

There are few reason why we should do the later approach. 

1. Making Monika to be able to "talk" to other monika is significantly more challenging compared to just simply execute other external process ( CURL ) using lots of SSH library.
2. CURL is much simpler, available on most (if not all) server in the world, and known to work.
3. CURL can do anything a HTTP client can do. 
4. Testing is much simpler, as we can easily simulate the remote calling (SSH) using CURL without the need to deploy anything remotely.
5. User do not have the burden of maintaining another remote app (e.g. remote monika server need to be maintained), If the remote curl is, less likely, to have problem, one can simply restart the server (or replace a simple VM).

**Reply by rmarszolek** [reply]
could you tell me where i find how to set up agent? I understand that I could in remote region set up agent and run probe on that agent . It could be usefull for checking if cloud services like traffic menager works properly , could you give me more details about agents ?

**Reply by sapiderman** [reply]
Hi @rmarszolek,

Please check the [installation docs here](https://monika.hyperjump.tech/quick-start). More guides can also be found on our [medium](https://medium.com/hyperjump-tech). How I would do it is probably using a [docker instance](https://monika.hyperjump.tech/tutorial/run-in-docker). Let us know if you get stuck installing/setting up Monika.

Regards.

**Comment by ariya** [comment]
An example config will be like the following:

```json
{
  "agents": [
    {
      "name": "Surabaya",
      "ip": "192.168.1.101",
      "username": "monika",
      "password": "trustno1"
    }
  ],
  "probes": [
    {
      "name": "Google",
      "description": "Check if Google is down or not",
      "interval": 60,
      "method": "GET",
      "timeout": 5,
      "requests": [
        {
          "url": "https://google.com"
        }
      ],
      "agents": [
        "local",
        "Surabaya"
      ]
    }
  ]
}
```

Here the probe named _Google_ will be executed on two places, _local_ (a special one) and _Surabaya_. The latter refers to the chosen agent name specified in the `agents` section.

The example illustrates the use username/password for authentication but obviously at some point we ought to support SSL certificate. Also for the time being, we can assume that the remote machine has cURL installed already.

**Reply by rmarszolek** [reply]
could you tell me where i find how to set up agent? I understand that I could in remote region set up agent and run probe on that agent . It could be usefull for checking if cloud services like traffic menager works properly , could you give me more details about agents ?

**Comment by ariawan** [comment]
Isn't this feature supposed to be the job of symon? not monika?

**Reply by ariya** [reply]
> Isn't this feature supposed to be the job of symon? not monika?

Based on some feedback, this feature is necessary to complement the Monika <-> Symon approach.

For some users, they want to get the benefit of getting their HTTP endpoints being probed from multiple locations, **without** Monika HQ. The reason is simplicity:

* Remote agent is trivial to setup and maintenance free: a simple VPS or VM (or the equivalent: EC2, GCE). There is also no dependency except cURL (almost de-facto standard).
* Such a remote agent can be provisioned, started, and tore down on-demand. This is important for folks already adopting infrastructure-as-code (Ansible, Terraform, etc).

**Reply by ariawan** [reply]
make sense, I agreed with this approach

**Comment by newm4n** [comment]
As POC, the following script should be imitated by monika probing.

```bash
#!/bin/bash

while getopts h:i:p: flag
do
    case "${flag}" in
        h) HOST=${OPTARG};;
        i) SSH_KEY=${OPTARG};;
        p) PROBE_TARGET=${OPTARG};;
    esac
done

if [ -z $HOST ] || [ -z $SSH_KEY ] || [ -z $PROBE_TARGET ]; then
    echo "Usage"
    echo "   -h SSH user@host target"
    echo "   -i Path to SSH private key or PEM"
    echo "   -p URL to proble"
    echo "Example"
    echo "    $0 -h user@myhost -i /path/to/key.pem -p https://check.com"
    exit 1
fi

if [ ! -r $SSH_KEY ]; then 
    echo "SSH Key not found"
    exit 1
fi

ssh -i $SSH_KEY $HOST bash -c "'
if [ -f arguments.txt ]; then
    rm arguments.txt
fi

# https://curl.se/docs/manpage.html

echo "{\\\\n" > arguments.txt
echo "\"monika_http_code\": %{http_code}\\\\n" >> arguments.txt
echo "\"monika_time_total\": %{time_total} seconds\\\\n" >> arguments.txt
echo "\"monika_size_request\": %{size_request} bytes\\\\n" >> arguments.txt
echo "\"monika_size_download\": %{size_download} bytes\\\\n" >> arguments.txt
echo "}\\\\n" >> arguments.txt

if [ -f detail.txt ]; then
    rm detail.txt
fi
if [ -f headers.txt ]; then
    rm headers.txt
fi
if [ -f output.txt ]; then
    rm output.txt
fi
curl -s --dump-header headers.txt -w "@arguments.txt" $PROBE_TARGET > detail.txt
echo "[response-detail]"
cat detail.txt | grep "monika"
echo "[response-headers]"
cat headers.txt
'"
```

save the script above as `remote.sh`. and execute as follows

```
$ sh remote.sh -h root@128.199.93.193 -i ~/.ssh/id_rsa_ferdinand_gmail -p https://github.com
Enter passphrase for key '/home/newm4n/.ssh/id_rsa_ferdinand_gmail': 
[response-detail]
monika_http_code: 200
monika_time_total: 1.166195 seconds
monika_size_request: 72 bytes
monika_size_download: 213087 bytes
[response-headers]
HTTP/2 200 
server: GitHub.com
date: Fri, 23 Apr 2021 03:24:19 GMT
content-type: text/html; charset=utf-8
vary: X-PJAX, Accept-Language, Accept-Encoding, Accept, X-Requested-With
permissions-policy: interest-cohort=()
etag: W/"5575dcadfce82522533e8a16b5819f73"
cache-control: max-age=0, private, must-revalidate
strict-transport-security: max-age=31536000; includeSubdomains; preload
x-frame-options: deny
x-content-type-options: nosniff
x-xss-protection: 0
referrer-policy: origin-when-cross-origin, strict-origin-when-cross-origin
expect-ct: max-age=2592000, report-uri="https://api.github.com/_private/browser/errors"
content-security-policy: default-src 'none'; base-uri 'self'; block-all-mixed-content; connect-src 'self' uploads.github.com www.githubstatus.com collector.githubapp.com api.github.com github-cloud.s3.amazonaws.com github-production-repository-file-5c1aeb.s3.amazonaws.com github-production-upload-manifest-file-7fdce7.s3.amazonaws.com github-production-user-asset-6210df.s3.amazonaws.com cdn.optimizely.com logx.optimizely.com/v1/events wss://alive.github.com github.githubassets.com; font-src github.githubassets.com; form-action 'self' github.com gist.github.com; frame-ancestors 'none'; frame-src render.githubusercontent.com; img-src 'self' data: github.githubassets.com identicons.github.com collector.githubapp.com github-cloud.s3.amazonaws.com secured-user-images.githubusercontent.com/ *.githubusercontent.com customer-stories-feed.github.com spotlights-feed.github.com; manifest-src 'self'; media-src github.githubassets.com; script-src github.githubassets.com; style-src 'unsafe-inline' github.githubassets.com; worker-src github.com/socket-worker-3f088aa2.js gist.github.com/socket-worker-3f088aa2.js
set-cookie: _gh_sess=y%2FUKbJcs2%2BgvAEXm5N2NMOgzDm4u5zqzGhXTSwAIRESb%2BjgA54%2FLt9zX0VhQkIJZdKHE2lPjP4DctnpY4Z5Crk7ALN5ZniD3CQhmzueQOzEG3QPwT%2FxCy5RLV83bjU%2FQ6I4%2F9vKmDQ%2FIGjUdHRrXAXqgLKUuYogFH4RC3d2CsHpgmw4U2OpSeveg3vgpkffvqGkar5UkOSwcmmqxFg6NzII3MGQBSM%2FngQ5vPLbxvB9bCh7ED0IfJDV6qJdEjNAkW%2FFWUglb6X4SwuC3zndKDw%3D%3D--76X73YXqLfb6S7%2Fw--ARuxG9Q0OtbwEHgNYCbgMw%3D%3D; Path=/; HttpOnly; Secure; SameSite=Lax
set-cookie: _octo=GH1.1.1715584545.1619148266; Path=/; Domain=github.com; Expires=Sat, 23 Apr 2022 03:24:26 GMT; Secure; SameSite=Lax
set-cookie: logged_in=no; Path=/; Domain=github.com; Expires=Sat, 23 Apr 2022 03:24:26 GMT; HttpOnly; Secure; SameSite=Lax
accept-ranges: bytes
x-github-request-id: CBCC:90ED:5548D8:5B7967:60823DE9
```

**Reply by newm4n** [reply]
This way, the http call is done by the destined host using CURL. The result of the call is directly captured from the caller (local) stdin.

**Reply by newm4n** [reply]
As shown, we captures the response code, response time, and response size. As well as all headers for chaining purpose.

**Reply by ariya** [reply]
Perhaps we also shall try the `--write-out` option described here: https://curl.se/docs/manpage.html#-w

**Reply by newm4n** [reply]
The script shown already demonstrate the `--write-out` (`-w`) option

```
curl -s --dump-header headers.txt -w "@arguments.txt" $PROBE_TARGET > detail.txt
```
